### PR TITLE
[Custom Metrics] Enabled support for logging of numerical metrics

### PR DESCRIPTION
**Pull request by MarkYHZhang**
Signed-off-by: Mark Zhang <mark.zhang@databricks.com>

## What changes are proposed in this pull request?

Added support for users to pass in custom metric functions into mlflow.evaluate. This PR is a part of more PRs to come for custom metrics support. Specifically, this PR focuses on enabling logging of numerical metrics generated by user's custom metric functions into MLflow.

## How is this patch tested?

Wrote unit tests located in mlflow/tests/models/test_default_evaluator.py

## Does this PR change the documentation?

Updates to the documentation will be done once the entire feature of custom metrics is complete.

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Enables MLflow tracking of custom metrics produced by user-provided metric functions.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by MarkYHZhang** [comment]
Hey @apurvakoti @dbczumar @jinzhang21 @WeichenXu123, can you all take a look? It seems like I don't have permission to add reviewers. So I'm tagging you all here. Thanks!

File: mlflow/models/evaluation/base.py
Comment (jinzhang21) [inline mlflow/models/evaluation/base.py:478]: Show what's required for the custom metric function signatures, and give an example on how it could be used (similar to what's discussed in the design discussion).

**Review by jinzhang21 — COMMENTED** [review]
Added some initial comment on the API. I'll review the rest later today.

File: mlflow/models/evaluation/default_evaluator.py
Comment (apurva-koti) [inline mlflow/models/evaluation/default_evaluator.py:592]: Can we add a TODO to log artifacts?

**Review by apurva-koti — COMMENTED** [review]
Some small comments, will do a full review again

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:266]: It would just return "<lambda>" as its "name".

File: mlflow/models/evaluation/base.py
Comment (jinzhang21) [reply mlflow/models/evaluation/base.py:472]: I'm generally not a big fan of allowing a mix of several types in a single function arg. i.e. for future extensibility, I'd rather prefer to have both `custom_metric_fns` and `custom_metrics` defined. On the other hand, allowing an input arg to have multiple types has been a tradition in MLflow (e.g. `dataset` can be a dataframe or ndarray). Purely for the sake of being internally consistent, I'm OK with using `custom_metrics` here.

File: mlflow/models/evaluation/default_evaluator.py
Comment (apurva-koti) [inline mlflow/models/evaluation/default_evaluator.py:297]:
```suggestion
        f"Custom metric '{custom_metric.__name__}' did not return an expected format."
```

**Review by apurva-koti — COMMENTED** [review]
Looks good! Let's use `[]` as the default argument for `custom_metrics` rather than `None`.

**Comment by MarkYHZhang** [comment]
> Looks good! Let's use `[]` as the default argument for `custom_metrics` rather than `None`.

I think we should avoid using mutable default argument in python due to issues described [here](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/). Although, I could initialize it to a list inside the constructor, but considering other default arguments like https://github.com/mlflow/mlflow/blob/2c0d13db6c677f1023a9a2a1776aca9e32aebe3d/mlflow/models/evaluation/base.py#L655
are not initialized, I think it might be good idea just to leave it as is.

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:297]: Is it bad practice to have long exception messages? Are users expected to refer to the docstring?

Comment (apurva-koti) [reply mlflow/models/evaluation/default_evaluator.py:297]: Just seemed like a duplication of information that could get annoying to maintain since it's not in an obvious place.
I think we can instead link them to the actual docs once they're published.

**Comment by apurva-koti** [comment]
Ah good catch @MarkYHZhang ! Let's keep it `None` then.

File: mlflow/models/evaluation/default_evaluator.py
Comment (apurva-koti) [inline mlflow/models/evaluation/default_evaluator.py:592]:
```suggestion
            if metric_results is not None:
                self.metrics.update(metric_results)
```

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:297]: Make sense!

File: tests/models/test_default_evaluator.py
Comment (apurva-koti) [inline tests/models/test_default_evaluator.py:553]: We should only assert on the exact substring that we know will probably persist, like `did not return in an expected format.` Maybe the fn name also.

**Review by apurva-koti — APPROVED** [review]
LGTM once all comments addressed!

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:845]: Can we use `.. code-block: python`, as in https://github.com/mlflow/mlflow/blob/2c0d13db6c677f1023a9a2a1776aca9e32aebe3d/mlflow/tracking/fluent.py#L208-L235?

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:262]: Can we add a short docstring for this internal method?

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:579]: in theory, it's possible for `self.y_pred` / `self.y` to be an object that gets modified during `_evaluate_custom_metric`. Embedding the object in a new pandas dataframe won't help in that case. Perhaps we want to use `deepcopy()` on `self.y_pred` and `self.y`.

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:831]: We have a branch cut for MLflow 1.24.0 coming up within the next 1-2 days. In case we want to include support for numerical metrics in 1.24.0 and follow up with artifacts in 1.25.0, can we remove this portion of the docstring since artifacts aren't yet supported?

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:283]: Shall we check the length of the tuple to make sure that it's `2`?

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:290]: Can we include `error_code=INVALID_PARAMETER_VALUE` here, indicating that the supplied `custom_metric` is invalid?

Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:290]: Can we link to the docs for expected input formats and / or provide a brief summary of the supported formats?

Comment (apurva-koti) [reply mlflow/models/evaluation/default_evaluator.py:290]: ^This existed but I asked @MarkYHZhang to take it out to avoid duplicating information that's hard to maintain (because this particular error isn't in an obvious place).
We will link to the docs once they're published

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:265]: Not all callables have a `__name__` attribute. Can we use `getattr(custom_metric, '__name__', <default>)` here and below instead? For <default> perhaps `"at index 0"`, `"at index 1"`, ... would be helpful, as it would tell the user the index of the custom metric function in the supplied `custom_metrics` list.

https://stackoverflow.com/a/33578357/11952869

**Review by dbczumar — COMMENTED** [review]
@MarkYHZhang This looks great! Left a few small comments - should be ready to merge soon!

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [reply mlflow/models/evaluation/default_evaluator.py:290]: @apurva-koti We should link to the docs in this PR, else initial users of the API won't see this helpful doc link in the error message. @MarkYHZhang can you include a link to the expected docs location that will exist once this PR is merged and docs are published? Happy to help.

Comment (apurva-koti) [reply mlflow/models/evaluation/default_evaluator.py:290]: Honestly it's probably fine to restore your earlier message describing the expected types @MarkYHZhang and then just replace it with a docs link when they're published. That way there's never an information gap

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:579]: Good catch!

File: mlflow/models/evaluation/base.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/base.py:831]: I think I should also modify the example to exclude artifacts support.

Comment (dbczumar) [reply mlflow/models/evaluation/base.py:831]: 👍

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [reply mlflow/models/evaluation/default_evaluator.py:290]: Can we include a link to the docs? We just need to link to `mlflow.evaluate()`. We already do this here: https://mlflow.org/docs/latest/models.html#model-evaluation.

File: mlflow/models/evaluation/base.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/base.py]: I have more written more detailed documentation along with example in the function docstring of the mlflow.evaluate function here https://github.com/mlflow/mlflow/blob/7b26a88d51d59ece67aca3a8db6a0140baf65822/mlflow/models/evaluation/base.py#L801-L839

**Review by dbczumar — COMMENTED** [review]
@MarkYHZhang Do we plan to add an example to the `examples` section or extend one of the existing examples? Can we do that as part of this PR?

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:817]: Can we add spaces between the list titles and the elements? The current rendering is probably not quite what we intend:

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:270]: Can we add a docstring for this parameter?

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:838]: Nit: Can we clean up the formatting here? Perhaps running this through `black` would be helpful.

**Review by dbczumar — APPROVED** [review]
LGTM when small docs nits have been addressed and example has been included. Thanks @MarkYHZhang ! Awesome stuff!

File: mlflow/models/evaluation/default_evaluator.py
Comment (harupy) [inline mlflow/models/evaluation/default_evaluator.py:276]:
```suggestion
    :return: A tuple of dictionaries. The first is a dictionary of metrics, the second is
             a dictionary of artifacts (which can be None if the custom metric function did
             not produce any).
```

Can we replace `null` with `None`? or are we using `null` intentionally?

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:287]: __validate_metrics_dict -> _validate_metrics

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:309]: The way the exception is raised is not informative. If I'm developing a custom metrics function, I have no way to know where it went wrong other than comparing against the documentation in mlflow.evaluate() very carefully. I added a few comments on where an exception should be raised above.

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:295]: If any of these condition is not true, raise in place with proper error message.

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:290]: If any of these condition is not true, raise in place with proper error message.

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:293]: __validate_artifacts_dict -> _validate_artifacts

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:606]: Why is user modifying self.metrics not concerning?

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:605]: This could be expensive. I recommend to only do it once outside of the loop.

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:279]: We raise on incorrect return format from `custom_metric` function. Why is None considered special here since it's not a correct format?

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:276]: Right, thanks!

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:606]: Right, that's a good point.

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:279]: Oh, because I was following this https://docs.google.com/document/d/1lC2PMENiT4G7o-QklRdeukGBON_hv9D47G1Q5GCt3GE/edit#bookmark=id.cod7z43dnwqy. I realized I didn't state this in the design document, should I just remove this behavior?

File: mlflow/models/evaluation/base.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/base.py:817]: Fixed! 
image

File: mlflow/models/evaluation/base.py
Comment (jinzhang21) [inline mlflow/models/evaluation/base.py:603]: Rm ","

File: mlflow/models/evaluation/base.py
Comment (jinzhang21) [inline mlflow/models/evaluation/base.py:669]: rm ","

File: mlflow/models/evaluation/default_evaluator.py
Comment (jinzhang21) [inline mlflow/models/evaluation/default_evaluator.py:626]: Will this step create another copy from copy_of_y_pred and copy_of_y (and a copy of dict is temporarily created)? Why not just moving this line outside of the loop?

File: tests/models/test_default_evaluator.py
Comment (jinzhang21) [inline tests/models/test_default_evaluator.py:507]: The test looks thorough, but it could be improved by using the parametrize method from pytest: https://docs.pytest.org/en/6.2.x/example/parametrize.html#parametrizing-conditional-raising

File: tests/models/test_default_evaluator.py
Comment (jinzhang21) [inline tests/models/test_default_evaluator.py:511]: Also add a test on lambda function. e.g. legal case: `custom_metrics=(lambda x: {'square': x**2},)` and illegal case: `custom_metrics=(lambda x: x,)`

File: mlflow/models/evaluation/base.py
Comment (harupy) [reply mlflow/models/evaluation/base.py:669]: This "," is automatically inserted by black. I'm curious why we need to remove it.

Comment (harupy) [reply mlflow/models/evaluation/base.py:669]: Is it for preventing this issue?
https://stackoverflow.com/questions/33350454/syntaxerror-with-passing-kwargs-and-trailing-comma

File: tests/models/test_default_evaluator.py
Comment (harupy) [inline tests/models/test_default_evaluator.py:610]:
```suggestion
    assert "pred_target_abs_diff" in res_artifacts_2
    assert res_artifacts_2[
        "pred_target_abs_diff"
    ].equals(np.abs(eval_df["prediction"] - eval_df["target"]))
```

Can we split this assertion to simplify the assertion error message?

### Example

```python
def test_1():
    res_artifacts_2 = {"example_dictionary_artifact": {"a": 1, "b": 3}}
    assert "example_dictionary_artifact" in res_artifacts_2
    assert res_artifacts_2["example_dictionary_artifact"] == {"a": 1, "b": 2}


"""
# pytest error message

E       AssertionError: assert {'a': 1, 'b': 3} == {'a': 1, 'b': 2}
E         Omitting 1 identical items, use -vv to show
E         Differing items:
E         {'b': 3} != {'b': 2}
E         Full diff:
E         - {'a': 1, 'b': 2}
E         ?               ^
E         + {'a': 1, 'b': 3}...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show
"""


def test_2():
    res_artifacts_2 = {"example_dictionary_artifact": {"a": 1, "b": 3}}
    assert "example_dictionary_artifact" in res_artifacts_2 and res_artifacts_2[
        "example_dictionary_artifact"
    ] == {"a": 1, "b": 2}


"""
# pytest error message

E       AssertionError: assert ('example_dictionary_artifact' in {'example_dictionary_artifact': {'a': 1, 'b': 3}} and {'a': 1, 'b': 3} == {'a': 1, 'b': 2}
E         Omitting 1 identical items, use -vv to show
E         Differing items:
E         {'b': 3} != {'b': 2}
E         Full diff:
E         - {'a': 1, 'b': 2}
E         ?               ^
E         + {'a': 1, 'b': 3}...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show)
"""

```

File: mlflow/models/evaluation/default_evaluator.py
Comment (MarkYHZhang) [reply mlflow/models/evaluation/default_evaluator.py:626]: Actually I just remembered that Corey brought up a good point [here](https://github.com/mlflow/mlflow/pull/5389#discussion_r813242477) that each custom metric function could modify `eval_df`. So it might be a good idea to perform deep copy on each. Namely reverting back to:
```python
eval_df = {"prediction": copy.deepcopy(self.y_pred), "target": copy.deepcopy(self.y)}
```

File: tests/models/test_default_evaluator.py
Comment (MarkYHZhang) [reply tests/models/test_default_evaluator.py:507]: Oh that's interesting, I didn't know about the parametrize method. But I wonder if I should use it here because this test only contain cases that raise exceptions. And seems like from the pytest documentation referenced above parametrize method is used in tests that contain cases that raise and don't raise exceptions. Perhaps did you mean I should rewrite my tests here so that they contain both exception-raising and exception-not-raising cases?

Comment (jinzhang21) [reply tests/models/test_default_evaluator.py:507]: No need to rewrite if it's difficult to factor them out. Leave them as is is fine.

File: mlflow/models/evaluation/base.py
Comment (jinzhang21) [reply mlflow/models/evaluation/base.py:669]: I recall there was some issue with it long time ago but don't remember exactly why. If black adds it, and it runs fine, I'm cool with it.

**Review by jinzhang21 — APPROVED** [review]

**Review by harupy — APPROVED** [review]

**Pull request merged**
